core/grants.go

```diff
@@ -0,0 +1,187 @@
+package core
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+
+	"chain/errors"
+	"chain/net/http/authz"
+	"chain/net/http/httpjson"
+)
+
+// an api-friendly representation of a grant
+type apiGrant struct {
+	GuardType string                 `json:"guard_type"`
+	GuardData map[string]interface{} `json:"guard_data"`
+	Policy    string                 `json:"policy"`
+	CreatedAt string                 `json:"created_at"`
+}
+
+// ErrMissingTokenID is returned when a token does not exist.
+var errMissingTokenID = errors.New("id does not exist")
+
+func (a *API) createGrant(ctx context.Context, x apiGrant) error {
+	id, ok := x.GuardData["id"]
+	if ok {
+		if strID, ok := id.(string); !ok || !a.accessTokens.Exists(ctx, strID) {
```

Comment:
This could avoid a level of nesting by using the unconditional map lookup (instead of the comma-ok form):

```go
if id, _ := x.GuardData["id"].(string); !a.accessTokens.Exists(ctx, id) {
	return errMissingTokenID
}
```